weldx/core.py

```diff
@@ -708,3 +711,557 @@ def units(self) -> pint.Unit:
+        self._units = expr_units
+        self._shape = expr_shape
+        self._obj = expr
+        self._variable_units = units
+        self._symbol_dims = dims
+
+    @staticmethod
+    def _eval_expr(expr, dims, units):
+        try:
+            scalar_params = {k: Q_(1, v) for k, v in units.items()}
+            result = expr.evaluate(**scalar_params)
+            expr_units = result.data.to_reduced_units().units
+        except pint.errors.DimensionalityError as e:
+            raise ValueError(
+                "Expression can not be evaluated due to a unit dimensionality error. "
+                "Ensure that the expressions parameters and the expected variable "
+                f"units are compatible. The original exception was:\n{e}"
+            )
+        except ValueError:
+            pass  # Error message will be generated by the next check
+
+        try:
+            # we evaluate twice with different array sizes because it might happen that
+            # a parameter uses the same dimension as a variable but the check still
+            # passes because the test input for the variable has the same array length.
+            # This case will be caught in the second evaluation.
+            for offset in range(2):
+                array_params = {
+                    k: xr.DataArray(Q_(range(i + 2 + offset), v), dims=dims[k])
+                    for i, (k, v) in enumerate(units.items())
+                }
+                expr.evaluate(**array_params)
+        except ValueError as e:
+            raise ValueError(
+                "During the evaluation of the expression mismatching array lengths' "
+                "were detected. Some possible causes are:\n"
+                "  - expression parameters that have already assigned coordinates to a "
+                "dimension that is also used as a variable\n"
+                "  - 2 free dimensions with identical names\n"
+                "  - 2 expression parameters that use the same dimension with "
+                "different number of values\n"
+                f"The original exception was:\n{e}"
+            )
+
+        # todo: shape should follow from dims of parameters and variables - Consider
+        #       removing shape for expressions since it does not really make sense. User
+        #       alternatives would be ndims and dims
+
+        shape = None
+        return expr_units, shape
+
+    @staticmethod
+    def _update_expression_params(params):
+        # todo
+        #  - enable usage of dicts for params (data, dims, coords)
+        #  - tuple should accept third element (coords)
+        for k, v in params.items():
+            if isinstance(v, tuple):
+                v = (Q_(v[0]), v[1])
+            elif isinstance(v, xr.DataArray):
+                if not isinstance(v.data, pint.Quantity):
+                    raise ValueError(f"Value for parameter {k} is not a quantity.")
+            else:
+                v = Q_(v)
+            params[k] = v
```

Comment:
As mentioned, docstring update will be done after the clean-up. 

The function modifies the dict in place. Therefore it does not need a return value. Its purpose is mainly to assure that every possible input format to the `MathematicalExpression` contains quantities. For example, the first branch keeps the tuple, only the first entry is converted (This can/will be rewritten a bit more direct)